File: Boggle/Elixir/test/boggle_test.exs
defmodule BoggleTest do
  use ExUnit.Case
  alias Boggle

  doctest Boggle

  # Test finds diagonal words on a board
  test "finds diagonal words on a board" do
    board = {
      {"c", "a", "t"},
      {"b", "t", "a"},
      {"d", "e", "k"}
    }
    words = ["cat", "bet"]

    result = Boggle.boggle(board, words)
    assert result == %{"bet" => [{1, 0}, {2, 1}, {1, 1}], "cat" => [{0, 0}, {0, 1}, {1, 1}]}
  end

  # Test recognizes words requiring backtracking
  test "recognizes words requiring backtracking" do
    board = {
      {"s", "t", "a", "r"},
      {"u", "r", "m", "s"},
      {"t", "a", "r", "t"},
      {"s", "t", "u", "n"}
    }
    words = ["start", "stun", "tart", "rum"]

    result = Boggle.boggle(board, words)
    assert result == %{"start" => [{3, 0}, {3, 1}, {2, 1}, {2, 2}, {2, 3}], "stun" => [{3, 0}, {3, 1}, {3, 2}, {3, 3}], "tart" => [{3, 1}, {2, 1}, {2, 2}, {2, 3}]}
  end

  # Test handles overlapping words correctly
  test "handles overlapping words correctly" do
    board = {
      {"s", "o", "s"},
      {"o", "a", "t"},
      {"m", "o", "m"}
    }
    words = ["so", "sat", "mom"]

    result = Boggle.boggle(board, words)
    assert result == %{"mom" => [{2, 2}, {2, 1}, {2, 0}], "sat" => [{0, 2}, {1, 1}, {1, 2}], "so" => [{0, 2}, {0, 1}]}
  end

  # Test finds horizontal and vertical words on a 2x2 board
  test "finds horizontal and vertical words on a 2x2 board" do
    board = {
      {"a", "b"},
      {"b", "a"}
    }
    words = ["ab", "ba"]

    result = Boggle.boggle(board, words)
    assert result == %{"ab" => [{1, 1}, {1, 0}], "ba" => [{1, 0}, {1, 1}]}
  end

  test "basic functionality on a 2x2 board" do
    words = ["eat", "tea", "ate"]
    board = {{"e", "a"}, {"t", "e"}}
    found = Boggle.boggle(board, words)
    
    # Validate the results
    assert validate_return_type(found), "Return type not correct"
    assert wordsLegal?(found, words), "Returned words not in list"
    assert wordsInBoard?(found, board), "Returned words not in board"

    # Calculate and log the score
    score = getScore(found)
    IO.puts "2x2 Basic Test Passed with score: #{score}"
  end

  test "all directions word search on a 4x4 board" do
    words = ["soup", "rope", "nerd", "den", "open", "pen"]
    board = {{"s", "o", "u", "p"}, {"r", "o", "p", "e"}, {"a", "b", "n", "d"}, {"n", "e", "r", "d"}}
    found = Boggle.boggle(board, words)
    
    # Validate the results
    assert validate_return_type(found), "Return type not correct"
    assert wordsLegal?(found, words), "Returned words not in list"
    assert wordsInBoard?(found, board), "Returned words not in board"

    # Calculate and log the score
    score = getScore(found)
    IO.puts "4x4 All Directions Test Passed with score: #{score}"
  end

  test "embedded board and word list on an 8x8 board" do
    # Predefined words
    words = [
      "connection", "leadership", "network", "programming", "algorithm",
      "function", "variable", "constant", "binary", "framework"
    ]

    # Predefined 8x8 board
    board = {
      {"c", "o", "n", "n", "e", "c", "t", "i"},
      {"o", "l", "e", "a", "d", "e", "r", "s"},
      {"n", "n", "e", "t", "w", "o", "r", "k"},
      {"p", "r", "o", "g", "r", "a", "m", "m"},
      {"a", "l", "g", "o", "r", "i", "t", "h"},
      {"f", "u", "n", "c", "t", "i", "o", "n"},
      {"v", "a", "r", "i", "a", "b", "l", "e"},
      {"c", "o", "n", "s", "t", "a", "n", "t"}
    }

    # Execute the Boggle game logic
    found = Boggle.boggle(board, words)
    
    # Validate the results
    assert validate_return_type(found), "Return type not correct"
    assert wordsLegal?(found, words), "Returned words not in list"
    assert wordsInBoard?(found, board), "Returned words not in board"

    # Calculate and log the score
    score = getScore(found)
    IO.puts "Embedded 8x8 Test Passed with score: #{score}"
  end

  test "embedded board and word list on a 16x16 board" do
    # Predefined words
    words = [
      "configuration", "establishment", "microprocessor", "telecommunication",
      "infrastructure", "superintendent", "multidimensional", "decentralization",
      "intercontinental", "philanthropist", "misinterpretation", "technologically",
      "incompatibilities", "disproportionate", "anthropological", "counterproductive"
    ]

    # Predefined 16x16 board
    board = {
      {"c", "o", "n", "f", "i", "g", "u", "r", "a", "t", "i", "o", "n", "a", "l", "x"},
      {"e", "s", "t", "a", "b", "l", "i", "s", "h", "m", "e", "n", "t", "r", "y", "z"},
      {"m", "i", "c", "r", "o", "p", "r", "o", "c", "e", "s", "s", "o", "r", "w", "k"},
      {"t", "e", "l", "e", "c", "o", "m", "m", "u", "n", "i", "c", "a", "t", "i", "o"},
      {"i", "n", "f", "r", "a", "s", "t", "r", "u", "c", "t", "u", "r", "e", "d", "p"},
      {"s", "u", "p", "e", "r", "i", "n", "t", "e", "n", "d", "e", "n", "t", "v", "q"},
      {"m", "u", "l", "t", "i", "d", "i", "m", "e", "n", "s", "i", "o", "n", "a", "l"},
      {"d", "e", "c", "e", "n", "t", "r", "a", "l", "i", "z", "a", "t", "i", "o", "n"},
      {"i", "n", "t", "e", "r", "c", "o", "n", "t", "i", "n", "e", "n", "t", "a", "l"},
      {"p", "h", "i", "l", "a", "n", "t", "h", "r", "o", "p", "i", "s", "t", "m", "c"},
      {"m", "i", "s", "i", "n", "t", "e", "r", "p", "r", "e", "t", "a", "t", "i", "o"},
      {"t", "e", "c", "h", "n", "o", "l", "o", "g", "i", "c", "a", "l", "l", "y", "u"},
      {"i", "n", "c", "o", "m", "p", "a", "t", "i", "b", "i", "l", "i", "t", "i", "e"},
      {"d", "i", "s", "p", "r", "o", "p", "o", "r", "t", "i", "o", "n", "a", "t", "e"},
      {"a", "n", "t", "h", "r", "o", "p", "o", "l", "o", "g", "i", "c", "a", "l", "f"},
      {"c", "o", "u", "n", "t", "e", "r", "p", "r", "o", "d", "u", "c", "t", "i", "v"}
    }

    # Execute the Boggle game logic
    found = Boggle.boggle(board, words)
    
    # Validate the results
    assert validate_return_type(found), "Return type not correct"
    assert wordsLegal?(found, words), "Returned words not in list"
    assert wordsInBoard?(found, board), "Returned words not in board"

    # Calculate and log the score
    score = getScore(found)
    IO.puts "Embedded 16x16 Test Passed with score: #{score}"
  end

  def validate_return_type(%{} = found) do
    found
    |> Map.values()
    |> Enum.any?(fn coords -> validate_coords_format(coords) end)
  end
  def validate_return_type(_), do: false
  defp validate_coords_format([{x, y} | _]) when is_integer(x) and is_integer(y), do: true
  defp validate_coords_format(_), do: false
  
  def wordsLegal?(found, words) do
    word_set = MapSet.new(words)
    found
    |> Map.keys()
    |> Enum.all?(fn word -> MapSet.member?(word_set, word) end)
  end
  
  def wordsInBoard?(found, board) do
    found
    |> Map.to_list()
    |> Enum.all?(fn {word, coords} -> 
      validate_word(board, word, coords) and validate_coords(coords)
    end)
  end
  defp validate_word(board, word, coords) do
    Enum.all?(Enum.zip(String.codepoints(word), coords), fn {char, {x, y}} -> 
      char == get_cell(board, x, y)
    end)
  end
  defp validate_coords(coords) do
    coords_unique?(coords) and coords_adjacent?(coords)
  end
  defp coords_unique?(coords), do: Enum.uniq(coords) == coords
  defp coords_adjacent?([{x, y} | _] = coords) do
    Enum.all?(Enum.zip(coords, tl(coords)), fn
      {{x1, y1}, {x2, y2}} -> abs(x1 - x2) <= 1 and abs(y1 - y2) <= 1
    end)
  end
  def get_cell(board, x, y), do: elem(elem(board, x), y)

  def getScore found do
    w_scores = {1, 2, 4, 6, 9, 12, 16, 20}  
    words = Map.keys found
    s = for w <- words, do: if( String.length(w) <= 8, do: elem(w_scores, String.length(w)-1), else: 20)
    Enum.sum s
  end
end

File: Boggle/Haskell/tests/BoggleTest.hs
module Main where
    import Test.HUnit
    import Data.Set (Set)
    import Data.List
    import Data.Array.IO
    import System.Random
    import Control.Monad
    import Data.Char (isSpace)
    import Data.Time.Clock.POSIX (getPOSIXTime)
    import Data.Char (isSpace)
    import Data.Time
    import qualified Boggle
    import qualified Data.Set as Set
    import qualified System.Exit as Exit

    main = do
        results <- runTestTT tests
        print results

    -- run tests with:
    --     cabal test

    boggle = Boggle.boggle
    wordsLegal :: [(String, [(Int, Int)])] -> [String] -> Bool
    wordsLegal found words =
        let wordSet = Set.fromList words
        in all (\(word, _) -> Set.member word wordSet) found

    wordsInBoard :: [ (String, [ (Int, Int) ] ) ] -> [[Char]] -> Bool
    wordsInBoard found board = 
        let res = map (\entry -> validateWord board entry) found 
        in  foldl (&&) True res

    wordCoordsOK found board =
        let res = map (\entry -> validateCoords entry) found 
        in  foldl (&&) True res

    validateWord :: [[Char]] -> ( [Char], [ (Int, Int) ] ) -> Bool
    validateWord board (word, coords) = 
        let zipped = zip word coords 
            res = map (\(ch, (x, y)) -> ch == (getCell board x y)) zipped
        in  foldl (&&) True res
        
    validateCoords (_, coords) = 
        let dup = length coords == Set.size (Set.fromList coords)
            xc = [ x | (x, _) <- coords ]
            x_ok = foldl (&&) True $ map (\(x1, x2) -> abs (x1-x2) <= 1) $ (zip xc $ tail xc) 
            yc = [ y | (_, y) <- coords ]
            y_ok = foldl (&&) True $ map (\(y1, y2) -> abs (y1-y2) <= 1) $ (zip yc $ tail yc) 
        in  dup && x_ok && y_ok

    getCell :: [[Char]] -> Int -> Int -> Char
    getCell board x y = (board !! x) !! y

    -- Calculates the score based on word lengths
    getScore :: [(String, [(Int, Int)])] -> Int
    getScore found =
        let wScores = [1, 2, 4, 6, 9, 12, 16, 20]
        in sum [if length w <= 8 then wScores !! (length w - 1) else 20 | (w, _) <- found]

    -- Shuffles a list in a random order
    shuffle :: [a] -> IO [a]
    shuffle xs = foldM removeRandom [] xs
        where removeRandom acc _ = do
                index <- getStdRandom $ randomR (0, length xs - 1)
                return $ (xs !! index) : acc

    -- Removes leading and trailing whitespace from a string
    trim :: String -> String
    trim = reverse . dropWhile isSpace . reverse . dropWhile isSpace
    
    tests :: Test
    tests = TestList [
        TestLabel "Test Basic Functionality on a 2x2 Board" testBasic2x2,
        TestLabel "Test All Directions on a 4x4 Board" testAllDirections4x4,
        TestLabel "Test Large Board 8x8" testEmbedded8x8Board,
        TestLabel "Test Massive Board 16x16" testEmbedded16x16Board, 
        TestLabel "Test Find Words on Board" testFindWordsOnBoard, 
        TestLabel "Test Backtracking and Overlapping" testBacktrackingAndOverlapping, 
        TestLabel "Test Diagonal Words" testDiagonalWordsOnBoard, 
        TestLabel "Test Overlapping Words" testHandlesOverlappingWords, 
        TestLabel "Test Find Words on 2x2 Board" testFindsWordsOn2x2Board
        ]

    testBasic2x2 :: Test
    testBasic2x2 = TestCase $ do
        let board = ["ea", "st"]
            words = ["eat", "tea", "ate"]
            found = boggle board words
            legalWords = wordsLegal found words
            inBoard = wordsInBoard found board
            coordsOK = wordCoordsOK found board
            score = getScore found
        assertBool "Check if all found words are legal" legalWords
        assertBool "Check if all words fit on the board" inBoard
        assertBool "Check if word coordinates are valid" coordsOK
        assertEqual "Check the scoring of found words" expectedScore score
        where expectedScore = 12  -- Assuming each word scores 4

    testAllDirections4x4 :: Test
    testAllDirections4x4 = TestCase $ do
        let board = ["soup", "rope", "abnd", "nerd"]
            words = ["soup", "rope", "nerd", "den", "open", "pen"]
            found = boggle board words
            legalWords = wordsLegal found words
            inBoard = wordsInBoard found board
            coordsOK = wordCoordsOK found board
            score = getScore found
            expectedScore = calculateExpectedScore ["soup", "rope", "nerd", "den"]

        -- Perform assertions to validate the conditions
        assertBool "Check if all found words are legal" legalWords
        assertBool "Check if all words fit on the board" inBoard
        assertBool "Check if word coordinates are valid" coordsOK
        assertEqual "Check the scoring of found words" expectedScore score
    
    testEmbedded8x8Board :: Test
    testEmbedded8x8Board = TestCase $ do
        let board = [
                "connecri", "oleaders", "nnetwork", "programming",
                "algorithm", "functions", "variable", "constants"
                ]
            words = ["connection", "leadership", "network", "programming", "algorithm",
                    "function", "variable", "constant", "binary", "framework"]
            found = boggle board words
            legalWords = wordsLegal found words
            inBoard = wordsInBoard found board
            coordsOK = wordCoordsOK found board
            score = getScore found
            expectedScore = calculateExpectedScore ["connection", "network", "algorithm", "variable"]

        assertBool "Check if all found words are legal" legalWords
        assertBool "Check if all words fit on the 8x8 board" inBoard
        assertBool "Check if word coordinates are valid on 8x8 board" coordsOK
        assertEqual "Check the scoring of found words on 8x8 board" expectedScore score

    testEmbedded16x16Board :: Test
    testEmbedded16x16Board = TestCase $ do
        let board = [
                "configuration", "establishment", "microprocessor", "telecommunication",
                "infrastructure", "superintendent", "multidimensional", "decentralization",
                "intercontinental", "philanthropist", "misinterpretation", "technologically",
                "incompatibilities", "disproportionate", "anthropological", "counterproductive"
                ]
            words = ["configuration", "establishment", "microprocessor", "telecommunication",
                    "infrastructure", "superintendent", "multidimensional", "decentralization",
                    "intercontinental", "philanthropist", "misinterpretation", "technologically",
                    "incompatibilities", "disproportionate", "anthropological", "counterproductive"]
            found = boggle board words
            legalWords = wordsLegal found words
            inBoard = wordsInBoard found board
            coordsOK = wordCoordsOK found board
            score = getScore found
            expectedScore = calculateExpectedScore ["configuration", "telecommunication", "multidimensional"]

        assertBool "Check if all found words are legal on 16x16 board" legalWords
        assertBool "Check if all words fit on the 16x16 board" inBoard
        assertBool "Check if word coordinates are valid on 16x16 board" coordsOK
        assertEqual "Check the scoring of found words on 16x16 board" expectedScore score

    testFindWordsOnBoard :: Test
    testFindWordsOnBoard = TestCase $ do
        let board = ["cat", "bta", "dek"]
            words = ["cat", "bet"]
            found = boggle board words
            expected = [("bet", [(1, 0), (2, 1), (1, 1)]), ("cat", [(0, 0), (0, 1), (1, 1)])]
        assertBool "finds diagonal words on a board" (wordsInBoard found board && wordsLegal found words)

    testBacktrackingAndOverlapping :: Test
    testBacktrackingAndOverlapping = TestCase $ do
        let board = ["star", "urms", "tart", "stun"]
            words = ["start", "stun", "tart", "rum"]
            found = boggle board words
            expected = [("start", [(0, 0), (0, 1), (0, 2), (1, 1), (2, 0)]),
                        ("stun", [(3, 0), (3, 1), (3, 2), (3, 3)]),
                        ("tart", [(2, 0), (2, 1), (2, 2), (2, 3)])]
        assertBool "recognizes words requiring backtracking" (wordsInBoard found board && wordsLegal found words)

    testDiagonalWordsOnBoard :: Test
    testDiagonalWordsOnBoard = TestCase $ do
        let board = ["cat", "bta", "dek"]
            words = ["cat", "bet"]
            found = boggle board words
            expected = [("bet", [(1, 0), (2, 1), (1, 1)]), ("cat", [(0, 0), (0, 1), (1, 1)])]
        assertBool "Diagonal and horizontal words test" (expected == found)

    testHandlesOverlappingWords :: Test
    testHandlesOverlappingWords = TestCase $ do
        let board = ["sos", "oat", "mom"]
            words = ["so", "sat", "mom"]
            found = boggle board words
            expected = [("mom", [(2, 0), (2, 1), (2, 2)]),
                        ("sat", [(0, 2), (1, 1), (1, 2)]),
                        ("so", [(0, 2), (0, 1)])]
        assertBool "Overlapping words test" (expected == found)

    testFindsWordsOn2x2Board :: Test
    testFindsWordsOn2x2Board = TestCase $ do
        let board = ["ab", "ba"]
            words = ["ab", "ba"]
            found = boggle board words
            expected = [("ab", [(0, 0), (0, 1)]), 
                        ("ba", [(1, 0), (1, 1)])]
        assertBool "2x2 board words test" (expected == found)

    -- Function to calculate the expected score based on a list of words
    calculateExpectedScore :: [String] -> Int
    calculateExpectedScore words = 
        let w_scores = [1, 2, 4, 6, 9, 12, 16, 20]
            scores = map (\w -> if length w <= 8 then w_scores !! (length w - 1) else 20) words
        in sum scores